core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1531,6 +1531,8 @@ class Partition(val topicPartition: TopicPartition,
             isrState = CommittedIsr(leaderAndIsr.isr.toSet)
             zkVersion = leaderAndIsr.zkVersion
             info(s"ISR updated to ${isrState.isr.mkString(",")} and version updated to [$zkVersion]")
+            leaderLogIfLocal.exists(leaderLog => maybeIncrementLeaderHW(leaderLog))
+
```

Comment:
Does maybeIncrementLeaderHW() (or its callees) take any locks that might lead to a deadlock if called from here?  (I'm not aware of any myself, this is just a sanity check.)

Reply:
That's definitely a good point. So I checked the maybeIncrementLeaderHW is using a lock on Log object, while shrinkISR is locked on updateLeaderAndIsr lock which should be on Partition level. I believe the 2 locks are not interact with each other, but honestly I don't know a good way to assert that 100% won't cause a deadlock.

Reply:
Good static analysis tools can spot such cases, but setting them up is a pain, and we don't have any such options ready to go, AFAIK. In lieu of that, sometimes lengthy stress-testing can work (or else place an upper limit on the probability of a deadlock happening); we could consider setting up something like that.

Reply:
Resolved via subsequent commits.